src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -78,42 +87,32 @@ partition_allocator::allocate_partition(
       clusterlog.trace,
       "allocating partition with constraints: {}",
       p_constraints);
+    uint16_t replicas_to_allocate = p_constraints.replication_factor
+                                    - not_changed_replicas.size();
     if (
-      p_constraints.replication_factor <= 0
-      || _state->available_nodes() < p_constraints.replication_factor) {
+      replicas_to_allocate <= 0
+      || _state->available_nodes() < replicas_to_allocate) {
         return errc::topic_invalid_replication_factor;
     }
 
     intermediate_allocation<model::broker_shard> replicas(
-      *_state, p_constraints.replication_factor, domain);
+      *_state, replicas_to_allocate, domain);
 
-    for (auto r = 0; r < p_constraints.replication_factor; ++r) {
-        auto effective_constraints = default_constraints(domain);
-        effective_constraints.hard_constraints.push_back(
-          ss::make_lw_shared<hard_constraint_evaluator>(
-            distinct_from(replicas.get())));
-
-        std::vector<model::broker_shard> current_replicas;
-        // rack-placement contraint
-        if (_enable_rack_awareness()) {
-            current_replicas = replicas.get();
-            current_replicas.insert(
-              current_replicas.end(),
-              not_changed_replicas.begin(),
-              not_changed_replicas.end());
-            effective_constraints.soft_constraints.push_back(
-              ss::make_lw_shared<soft_constraint_evaluator>(
-                distinct_rack_preferred(current_replicas, *_state)));
-        }
+    std::vector<model::broker_shard> all_replicas = not_changed_replicas;
```

Comment:
nit: we could probably just use `replicas.get()` here and get rid of `all_replicas`

Reply:
problem with using `replicas` here is that it is an `intermediate_reallocation` which is a RAII wrapper that reverts reallocation

Reply:
Not sure I understand - all that we need `all_replicas` for is to hold the set of current replicas and to pass it to `_allocation_strategy.allocate_replica` (and the parameter reference does not escape the call AFAICS). And while we are inside this function, `all_replicas == replicas.get()`. Am I missing something?

Reply:
I was thinking about this but then it came to me why we need `all_replicas` the `intermediate_allocation<model::broker_shard> replicas` is an intermediate allocation i.e. it only holds newly allocated nodes. In case of a failure or an exception it will revert back the ongoing allocation. For this to be possible it must only contain new allocations not the one that are current. Hence the need to another vector with `all_replicas` if we want to reallocate only one replica for partition the `replicas` should only contain that one replica for the purpose of reverting, while `all_replicas` must contain all of them both newly allocated and the old one for the constraints like `distinct_rack`.

Reply:
In that case this patch introduces a behavioral change, right? Because earlier we passed only new replicas  to all the constraints except distinct_rack() but with the patch, we pass all_replicas for all constraints or am I missing something?

Reply:
there is no behavioral change as we still use the same approach. With the new interface a constraint can access all replicas if it needs to.

Reply:
got it, I think I got confused, actually `all_replicas != replicas.get()`

Reply:
> there is no behavioral change as we still use the same approach. With the new interface a constraint can access all replicas if it needs to.

Just to clarify what I meant. Without this patch, IIUC, `distinct_from` (L94) was computing distinctness over new replicas but with the patch it computes it over all_replicas.